velox/common/memory/SharedArbitrator.h

```diff
@@ -231,32 +321,41 @@ class SharedArbitrator : public memory::MemoryArbitrator {
   // shrink.
   int64_t reclaimableFreeCapacity(const MemoryPool& pool) const;
 
-  // Returns the used memory capacity that can be reclaimed from 'pool' by disk
-  // spill.
+  // Returns the used memory capacity that can be reclaimed from 'pool' by
+  // disk spill.
   int64_t reclaimableUsedCapacity(const MemoryPool& pool) const;
 
   // Returns the minimal amount of memory capacity to grow for 'pool' to have
   // the reserved capacity as specified by 'memoryPoolReservedCapacity_'.
   int64_t minGrowCapacity(const MemoryPool& pool) const;
 
+  void updateArbitrationRequestStats();
+  void updateArbitrationFailureStats();
+
+  // Lock used to protect the arbitrator state.
   mutable std::mutex mutex_;
   tsan_atomic<uint64_t> freeReservedCapacity_{0};
   tsan_atomic<uint64_t> freeNonReservedCapacity_{0};
-  // Indicates if there is a running arbitration request or not.
-  bool running_{false};
-
-  // The promises of the arbitration requests waiting for the serialized
-  // execution.
-  std::vector<ContinuePromise> waitPromises_;
-
-  tsan_atomic<uint64_t> numRequests_{0};
-  std::atomic<uint64_t> numSucceeded_{0};
+  // Contains the arbitration running queues with one per each query memory
+  // pool.
+  std::unordered_map<MemoryPool*, std::unique_ptr<ArbitrationQueue>>
+      arbitrationQueues_;
+
+  // R/W lock used to control local and global arbitration runs. A local
+  // arbitration run needs to hold a shared lock while the latter needs to hold
+  // an exclusive lock. Hence, multiple local arbitration runs from different
+  // query memory pools can run in parallel but the global ones has to run with
+  // one at a time.
+  mutable std::shared_mutex arbitrationLock_;
+
+  std::atomic_uint64_t numRequests_{0};
+  std::atomic_uint32_t numRunning_{0};
   tsan_atomic<uint64_t> numAborted_{0};
-  tsan_atomic<uint64_t> numFailures_{0};
-  tsan_atomic<uint64_t> queueTimeUs_{0};
+  std::atomic_uint64_t numFailures_{0};
+  std::atomic_uint64_t waitTimeUs_{0};
   tsan_atomic<uint64_t> arbitrationTimeUs_{0};
-  tsan_atomic<uint64_t> numShrunkBytes_{0};
-  tsan_atomic<uint64_t> numReclaimedBytes_{0};
+  tsan_atomic<uint64_t> reclaimedFreeBytes_{0};
+  tsan_atomic<uint64_t> reclaimedUsedBytes_{0};
   tsan_atomic<uint64_t> reclaimTimeUs_{0};
   tsan_atomic<uint64_t> numNonReclaimableAttempts_{0};
   tsan_atomic<uint64_t> numReserves_{0};
```

Comment:
Do we want to separately expose local arbitration stats as well? That might be valuable?

Reply:
Since this is a newly added major feature, we might want to expose a bit more finer granularity metrics for better monitoring?

Reply:
We have runtime metrics for these. We can add fine-grained logging in followup.